### ISSUE TITLE
Gix-SQL: please drop duplicate calls to SetSQLParams for reused parameters

### ISSUE THREAD

**Issue by GitMensch** [mridoni/gix#36]
Given the sample

```cobol
           EXEC SQL AT :DBS
              DECLARE CSKEY08 CURSOR FOR
                 SELECT FLD1, FLD2 FROM TAB
                 WHERE
                    KEY08 >= (
                        CASE WHEN :AMNT< 0.000 THEN
                           'A' || TO_CHAR(999999999999999.999 +
                                          :AMNT)
                        ELSE
                           'B' || TO_CHAR(0.000 +
                                          :AMNT)
                        END
                             )
                 ORDER BY KEY08 ASC
           END-EXEC.
```

the statement generated by the old OCESQL is
```sql
SELECT FLD1, FLD2 FROM TAB WHERE KEY08 >= ( CASE WHEN $1 < 0.000 THEN 'A' || TO_CHAR(999999999999999.999 + $2) ELSE 'B' || TO_CHAR(0.000 + $3) END ) ORDER BY KEY08 ASC
```
and COBOL:
```cobol
OCESQL     CALL "OCESQLStartSQL"
OCESQL     END-CALL
OCESQL     CALL "OCESQLSetSQLParams" USING
OCESQL          BY VALUE 9
OCESQL          BY VALUE 18
OCESQL          BY VALUE -3
OCESQL          BY REFERENCE AMNT
OCESQL     END-CALL
OCESQL     CALL "OCESQLSetSQLParams" USING
OCESQL          BY VALUE 9
OCESQL          BY VALUE 18
OCESQL          BY VALUE -3
OCESQL          BY REFERENCE AMNT
OCESQL     END-CALL
OCESQL     CALL "OCESQLSetSQLParams" USING
OCESQL          BY VALUE 9
OCESQL          BY VALUE 18
OCESQL          BY VALUE -3
OCESQL          BY REFERENCE AMNT
OCESQL     END-CALL
OCESQL     CALL "OCESQLIDCursorDeclareParams" USING
OCESQL          BY REFERENCE SQLCA
OCESQL          BY REFERENCE DBS
OCESQL          BY VALUE 10
OCESQL          BY REFERENCE "PROG_CSKEY08" & x"00"
OCESQL          BY REFERENCE SQ0030
OCESQL          BY VALUE 3
OCESQL     END-CALL
OCESQL     CALL "OCESQLEndSQL"
OCESQL     END-CALL.
```

While the generation _should_ be something along
```sql
SELECT FLD1, FLD2 FROM TAB WHERE KEY08 >= ( CASE WHEN $1 < 0.000 THEN 'A' || TO_CHAR(999999999999999.999 + $1) ELSE 'B' || TO_CHAR(0.000 + $1) END ) ORDER BY KEY08 ASC
```
and
```cobol
OCESQL     CALL "OCESQLStartSQL"
OCESQL     END-CALL
OCESQL     CALL "OCESQLSetSQLParams" USING
OCESQL          BY VALUE 9
OCESQL          BY VALUE 18
OCESQL          BY VALUE -3
OCESQL          BY REFERENCE AMNT
OCESQL     END-CALL
OCESQL     CALL "OCESQLIDCursorDeclareParams" USING
OCESQL          BY REFERENCE SQLCA
OCESQL          BY REFERENCE DBS
OCESQL          BY VALUE 10
OCESQL          BY REFERENCE "PROG_CSKEY08" & x"00"
OCESQL          BY REFERENCE SQ0030
OCESQL          BY VALUE 1
OCESQL     END-CALL
OCESQL     CALL "OCESQLEndSQL"
OCESQL     END-CALL.
```

Can you please ensure that each parameter is only bind one time - even if it is referenced multiple times?
Maybe that's already taken care of?

I would like to see it solved in the next release - which will be the first time I can actually _check_ things because of the separate Gix-SQL package without the need for QT-anything (I _hope_ this can be compiled with plain `make`, possibly with a configure script).

**Comment by mridoni** [comment]
This would require a change in the way parameters are handled: currently there is no parameter "index" ($1, $2, etc.) passed to the backend (we would need the index to reuse the parameter definition) and this would be useless anyway when using anonymous parameters. In this latter case we could actually use the reference field, but it would make things more complicated. I'll leave it here since any idea for an optimization is welcome, but it's not going to gain high priority soon.

**Comment by GitMensch** [comment]
Isn't the index passed via the generated SQL statement (`SQ0030` above) - or did you change that in GixSQL?
If it is passed then only the preparser would need to keep the list within the same statement.

**Comment by mridoni** [comment]
No, that's the variable the parameter points to, I was talking about the index for the parameter in the SQL statement ($1, $2, etc.) that can be reused in the statement itself when the DBMS accepts named parameters (BTW, it is not necessarily a number) as in 

      insert into mytbl (a, b, c, d) values ($1, $1, $2, $2) 

For instance ODBC only accepts anonymous parameters (except for stored procedures) so you always have to bind a single parameter multiple times, even if the underlying variable is the same.

Anyway, while the suggestion is welcome, this is a minor optimization: the preprocessed COBOL code would become only marginally shorter and the efficiency in binding less parameters would be easily offset by having to keep track of them, while making parsing more difficult and error-prone.

**Comment by GitMensch** [comment]
> the efficiency in binding less parameters would be easily offset by having to keep track of them

Hm, likely depends on how often the same parameter is used in the same statement.
But there wouldn't be no need to keep track of them during runtime because with a changed preparser they only would be included once each in the statement (as above) and the `nParams` would also be reduced (and therefore the calls to a bunch of other functions, not only `SQLBindParameter()`).

... and, of course, it would make the generated code look more similar to the one from Pro*COB and [esqloc](https://sourceforge.net/p/gnucobol/contrib/HEAD/tree/trunk/esql/)...

> For instance ODBC only accepts anonymous parameters (except for stored procedures) so you always have to bind a single parameter multiple times, even if the underlying variable is the same.

That seems weird. Doesn't `SQLSetDescField()` allow you to explicit name the parameter? In this case it could be named P1, P2, ... and the generated code be changed to

```sql
  insert into mytbl (a, b, c, d) values (@P1, @P1, @P2, @P2) 
```
no?

**Comment by mridoni** [comment]
As far as I can tell this method only works for stored procedures. The Microsoft documentation for ODBC (but I guess it's part of the standard) specifically states: "In ODBC, named parameters are used only in calls to stored procedures and cannot be used in other SQL statements."

https://docs.microsoft.com/en-us/sql/odbc/reference/develop-app/binding-parameters-by-name-named-parameters?view=sql-server-ver15
